# Business Understanding

### Predict house prices for negotiation with house sellers

# Data Understanding

### Load Data

### Check the number of rows and columns in datasets

In [ ]:
print('train:', train.shape)
print('test:', test.shape)

train: (1460, 81)
test: (1459, 80)


### Skim the data in datasets

In [ ]:
train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [ ]:
test.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

### Data Types

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [ ]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1459 non-null   int64  
 1   MSSubClass     1459 non-null   int64  
 2   MSZoning       1455 non-null   object 
 3   LotFrontage    1232 non-null   float64
 4   LotArea        1459 non-null   int64  
 5   Street         1459 non-null   object 
 6   Alley          107 non-null    object 
 7   LotShape       1459 non-null   object 
 8   LandContour    1459 non-null   object 
 9   Utilities      1457 non-null   object 
 10  LotConfig      1459 non-null   object 
 11  LandSlope      1459 non-null   object 
 12  Neighborhood   1459 non-null   object 
 13  Condition1     1459 non-null   object 
 14  Condition2     1459 non-null   object 
 15  BldgType       1459 non-null   object 
 16  HouseStyle     1459 non-null   object 
 17  OverallQual    1459 non-null   int64  
 18  OverallC

### Data information

##### Apparently, all columns in the datasets seem informative variables based on the data explanation.
##### Link: [The explanation about the columns of the datasets for this project](https://www.kaggle.com/competitions/house-prices-advanced-regression-techniques/data)

# Data Preparation

### Convert categorical data into numeric data

In [ ]:
train = pd.get_dummies(train)
test = pd.get_dummies(test)

##### And the next thing to do is checking whether the new columns in train and test dataset are the same or not.

In [ ]:
print('the number of columns in train dataset is', len(train.columns))
print('the number of columns in test dataset is', len(test.columns))

the number of columns in train dataset is 290
the number of columns in test dataset is 271


##### The number of columns in train and test dataset are different, so leave only the same columns in train dataset.

In [ ]:
len(train[test.columns].columns)

271

##### The result above is the same as the number of columns in test dataset, so let's change train dataset to dataset having only the same columns. 
##### Warning! The dependent variable does not exist in test dataset, so let's add it into new train dataset.

In [ ]:
dependent_variable = train.SalePrice

In [ ]:
train = train[test.columns]
train['SalePrice'] = dependent_variable

### Set Index

In [ ]:
train = train.set_index('Id')
test = test.set_index('Id')

### Collinearity

In [ ]:
condition = (train.corr() < -0.7) | (train.corr() > 0.7) & (train.corr() != 1)
correlation = train.drop('SalePrice', axis=1).corr()[condition].dropna(how='all', axis=0)
collinearity = correlation[correlation.index]

In [ ]:
collinearity

MSSubClass  YearBuilt  BsmtFinSF2  TotalBsmtSF  \
MSSubClass                    NaN        NaN         NaN          NaN   
YearBuilt                     NaN        NaN         NaN          NaN   
BsmtFinSF2                    NaN        NaN         NaN          NaN   
TotalBsmtSF                   NaN        NaN         NaN          NaN   
1stFlrSF                      NaN        NaN         NaN      0.81953   
...                           ...        ...         ...          ...   
PavedDrive_Y                  NaN        NaN         NaN          NaN   
MiscFeature_Gar2              NaN        NaN         NaN          NaN   
SaleType_New                  NaN        NaN         NaN          NaN   
SaleType_WD                   NaN        NaN         NaN          NaN   
SaleCondition_Partial         NaN        NaN         NaN          NaN   

                       1stFlrSF  2ndFlrSF  GrLivArea  TotRmsAbvGrd  \
MSSubClass                  NaN       NaN        NaN           NaN   
YearBuilt                   NaN       NaN        NaN           NaN   
BsmtFinSF2                  NaN       NaN        NaN           NaN   
TotalBsmtSF             0.81953       NaN        NaN           NaN   
1stFlrSF                    NaN       NaN        NaN           NaN   
...                         ...       ...        ...           ...   
PavedDrive_Y                NaN       NaN        NaN           NaN   
MiscFeature_Gar2            NaN       NaN        NaN           NaN   
SaleType_New                NaN       NaN        NaN           NaN   
SaleType_WD                 NaN       NaN        NaN           NaN   
SaleCondition_Partial       NaN       NaN        NaN           NaN   

                       GarageYrBlt  GarageCars  ...  GarageType_Attchd  \
MSSubClass                     NaN         NaN  ...                NaN   
YearBuilt                 0.825667         NaN  ...                NaN   
BsmtFinSF2                     NaN         NaN  ...                NaN   
TotalBsmtSF                    NaN         NaN  ...                NaN   
1stFlrSF                       NaN         NaN  ...                NaN   
...                            ...         ...  ...                ...   
PavedDrive_Y                   NaN         NaN  ...                NaN   
MiscFeature_Gar2               NaN         NaN  ...                NaN   
SaleType_New                   NaN         NaN  ...                NaN   
SaleType_WD                    NaN         NaN  ...                NaN   
SaleCondition_Partial          NaN         NaN  ...                NaN   

                       GarageType_Detchd  GarageQual_TA  GarageCond_TA  \
MSSubClass                           NaN            NaN            NaN   
YearBuilt                            NaN            NaN            NaN   
BsmtFinSF2                           NaN            NaN            NaN   
TotalBsmtSF                          NaN            NaN            NaN   
1stFlrSF                             NaN            NaN            NaN   
...                                  ...            ...            ...   
PavedDrive_Y                         NaN            NaN            NaN   
MiscFeature_Gar2                     NaN            NaN            NaN   
SaleType_New                         NaN            NaN            NaN   
SaleType_WD                          NaN            NaN            NaN   
SaleCondition_Partial                NaN            NaN            NaN   

                       PavedDrive_N  PavedDrive_Y  MiscFeature_Gar2  \
MSSubClass                      NaN           NaN               NaN   
YearBuilt                       NaN           NaN               NaN   
BsmtFinSF2                      NaN           NaN               NaN   
TotalBsmtSF                     NaN           NaN               NaN   
1stFlrSF                        NaN           NaN               NaN   
...                             ...           ...               ...   
PavedDrive_Y              -0.8

##### Choose one column going to be removed among two columns, which have collinearity with each other

In [ ]:
x = list(collinearity.columns)
for i in x:
    x.remove(collinearity[i].dropna().index[0])
x

['MSSubClass',
 'YearBuilt',
 'BsmtFinSF2',
 'TotalBsmtSF',
 'GrLivArea',
 'GarageCars',
 'MiscVal',
 'MSZoning_FV',
 'MSZoning_RL',
 'Street_Grvl',
 'LotShape_IR1',
 'LotConfig_Corner',
 'LandSlope_Gtl',
 'HouseStyle_2Story',
 'RoofStyle_Gable',
 'RoofMatl_Tar&Grv',
 'Exterior1st_CemntBd',
 'Exterior1st_HdBoard',
 'Exterior1st_MetalSd',
 'Exterior1st_Plywood',
 'Exterior1st_Stucco',
 'Exterior1st_VinylSd',
 'Exterior1st_Wd Sdng',
 'Exterior2nd_AsbShng',
 'MasVnrType_None',
 'ExterQual_TA',
 'ExterCond_TA',
 'Foundation_PConc',
 'BsmtQual_TA',
 'Heating_GasW',
 'CentralAir_Y',
 'Electrical_SBrkr',
 'KitchenQual_TA',
 'GarageType_Detchd',
 'GarageCond_TA',
 'PavedDrive_Y',
 'SaleType_WD',
 'SaleCondition_Partial']

### Mulitcollinearity among columns having collinearity

In [ ]:
collinearity[train[x].corr()[condition].isnull() == False].dropna(how='all')

MSSubClass  YearBuilt  BsmtFinSF2  TotalBsmtSF  \
SaleType_WD                   NaN        NaN         NaN          NaN   
SaleCondition_Partial         NaN        NaN         NaN          NaN   

                       1stFlrSF  2ndFlrSF  GrLivArea  TotRmsAbvGrd  \
SaleType_WD                 NaN       NaN        NaN           NaN   
SaleCondition_Partial       NaN       NaN        NaN           NaN   

                       GarageYrBlt  GarageCars  ...  GarageType_Attchd  \
SaleType_WD                    NaN         NaN  ...                NaN   
SaleCondition_Partial          NaN         NaN  ...                NaN   

                       GarageType_Detchd  GarageQual_TA  GarageCond_TA  \
SaleType_WD                          NaN            NaN            NaN   
SaleCondition_Partial                NaN            NaN            NaN   

                       PavedDrive_N  PavedDrive_Y  MiscFeature_Gar2  \
SaleType_WD                     NaN           NaN               NaN   
SaleCondition_Partial           NaN           NaN               NaN   

                       SaleType_New  SaleType_WD  SaleCondition_Partial  
SaleType_WD                     NaN          NaN              -0.769559  
SaleCondition_Partial           NaN    -0.769559                    NaN  

[2 rows x 76 columns]

In [ ]:
x.remove('SaleType_WD')

##### Remove the columns being decided to be discarded above

In [ ]:
train = train.drop(x, axis=1)
test = test.drop(x, axis=1)

### Missing Values

##### Make sure to remove the same columns in test dataset when we try to remove columns having a lot of missing values in train dataset, and vice versa.

In [ ]:
(train.isnull().sum() / train.shape[0])[train.columns[train.isnull().sum() > 0]]

LotFrontage    0.177397
MasVnrArea     0.005479
GarageYrBlt    0.055479
dtype: float64

##### 'LotFrontage', 'GarageYrBlt' must be discarded.

In [ ]:
train = train.drop(['LotFrontage', 'GarageYrBlt'], axis=1)
test = test.drop(['LotFrontage', 'GarageYrBlt'], axis=1)

##### And only the missing values in 'MasVnrArea' will be removed, because the proportion of missing values in the rest of columns is not high.

In [ ]:
train[train.MasVnrArea.isnull() == True]

LotArea  OverallQual  OverallCond  YearRemodAdd  MasVnrArea  BsmtFinSF1  \
Id                                                                              
235      7851            6            5          2002         NaN         625   
530     32668            6            3          1975         NaN        1219   
651      8125            7            6          2007         NaN           0   
937     10083            7            5          2003         NaN         833   
974     11639            7            5          2008         NaN           0   
978      4274            7            5          2007         NaN        1106   
1244    13891           10            5          2006         NaN        1386   
1279     9473            8            5          2002         NaN         804   

      BsmtUnfSF  1stFlrSF  2ndFlrSF  LowQualFinSF  ...  SaleType_ConLw  \
Id                                                 ...                   
235         235       860      1100             0  ...               0   
530         816      2515         0             0  ...               0   
651         813       822       843             0  ...               0   
937         343      1200         0             0  ...               0   
974        1428      1428         0             0  ...               0   
978         135      1241         0             0  ...               0   
1244        690      2076         0             0  ...               0   
1279        324      1128       903             0  ...               0   

      SaleType_New  SaleType_Oth  SaleType_WD  SaleCondition_Abnorml  \
Id                                                                     
235              0             0            1                      0   
530              0             0            1                      0   
651              0             0            1                      0   
937              0             0            1                      0   
974              1             0            0                      0   
978              1             0            0                      0   
1244             1             0            0                      0   
1279             0             0            1                      0   

      SaleCondition_AdjLand  SaleCondition_Alloca  SaleCondition_Family  \
Id                                                                        
235                       0                     0                     0   
530                       0                     1                     0   
651                       0                     0                     0   
937                       0                     0                     0   
974                       0                     0                     0   
978                       0                     0                     0   
1244                      0                     0                     0   
1279                      0                     0                     0   

      SaleCondition_Normal  SalePrice  
Id                                     
235                      1     216500  
530                      0     200624  
651                      1     205950  
937                      1     184900  
974                      0     182000  
978                      0     199900  
1244                     0     465000  
1279                     1     237000  

[8 rows x 232 columns]

In [ ]:
train = train.drop(train[train.MasVnrArea.isnull() == True].index)

##### Make sure to replace missing values in test datasets, not to remove them.
##### If we need to remove the missing values, then need to discard the column having the values, because submission file should be a CSV file with 1459 rows. (Let's think this requirement as specific order from our customer who wants this model in real life.)

In [ ]:
(test.isnull().sum() / test.shape[0])[test.columns[test.isnull().sum() > 0]]

MasVnrArea      0.010281
BsmtFinSF1      0.000685
BsmtUnfSF       0.000685
BsmtFullBath    0.001371
BsmtHalfBath    0.001371
GarageArea      0.000685
dtype: float64

##### Fortunately, there are not columnes, which have high proportion of missing values in test dataset.

In [ ]:
null_columns = list((test.isnull().sum() / test.shape[0])[test.columns[test.isnull().sum() > 0]].index)

##### Let's measure the correlation between dependent variable and columns having missing values.
##### If the correlation is not high, the columns (independent variables) are useless for prediction. So it is proper to remove those columns.

In [ ]:
train.corr()['SalePrice'][null_columns]

MasVnrArea      0.477493
BsmtFinSF1      0.383977
BsmtUnfSF       0.215740
BsmtFullBath    0.225027
BsmtHalfBath   -0.015993
GarageArea      0.622492
Name: SalePrice, dtype: float64

##### Except GarageArea, the rest of columns has low correlation with 'SalePrice'.
##### Thus, let's remove them.

In [ ]:
train = train.drop(['MasVnrArea', 'BsmtFinSF1', 'BsmtUnfSF', 'BsmtFullBath', 'BsmtHalfBath'], axis=1)
test = test.drop(['MasVnrArea', 'BsmtFinSF1', 'BsmtUnfSF', 'BsmtFullBath', 'BsmtHalfBath'], axis=1)

##### And fill the missing value in GarageArea.

In [ ]:
test[test.GarageArea.isnull() == True]

LotArea  OverallQual  OverallCond  YearRemodAdd  1stFlrSF  2ndFlrSF  \
Id                                                                          
2577     9060            5            6          1999       942       886   

      LowQualFinSF  FullBath  HalfBath  BedroomAbvGr  ...  SaleType_ConLI  \
Id                                                    ...                   
2577             0         2         0             3  ...               0   

      SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
Id                                                              
2577               0             0             0            1   

      SaleCondition_Abnorml  SaleCondition_AdjLand  SaleCondition_Alloca  \
Id                                                                         
2577                      0                      0                     1   

      SaleCondition_Family  SaleCondition_Normal  
Id                                                
2577                     0                     0  

[1 rows x 226 columns]

##### The number of missing value is just one, so let's fill the value by the mean of the variable.

In [ ]:
test.GarageArea = test.GarageArea.fillna(round(test.GarageArea.mean(),0))

### Set Dependent Variable and Independent Variables

In [ ]:
X = train.drop('SalePrice', axis=1)
y = train.SalePrice

### Standardization

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X[:] = scaler.fit_transform(X)
test[:] = scaler.fit_transform(test)

### Split Dataset into Train, Validation, Test datasets

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.20, random_state=2023)
X_test = test

# Modeling

### Ridge Regression

In [ ]:
from sklearn.linear_model import Ridge
r = Ridge(alpha=100)
r.fit(X_train, y_train)
y_pred_train_r = r.predict(X_train)
y_pred_val_r = r.predict(X_val)

### Lasso Regression

In [ ]:
from sklearn.linear_model import Lasso
l = Lasso(alpha=100)
l.fit(X_train, y_train)
y_pred_train_l = l.predict(X_train)
y_pred_val_l = l.predict(X_val)

### Support Vector Machine Regression

In [ ]:
from sklearn.svm import SVR
svr = SVR()
svr.fit(X_train, y_train)
y_pred_train_svr = svr.predict(X_train)
y_pred_val_svr = svr.predict(X_val)

### Random Forest Regression

In [ ]:
from sklearn.ensemble import RandomForestRegressor
rfr = RandomForestRegressor(n_estimators=500, max_features=int(len(train.columns)*0.7))
rfr.fit(X_train, y_train)
y_pred_train_rfr = rfr.predict(X_train)
y_pred_val_rfr = rfr.predict(X_val)

# Evaluation

In [ ]:
from sklearn.metrics import mean_squared_error
import math

### RMSE_Train

In [ ]:
print('Ridge Regression:', math.sqrt(mean_squared_error(y_train, y_pred_train_r)))
print('Lasso Regression:', math.sqrt(mean_squared_error(y_train, y_pred_train_l)))
print('Support Vector Machine Regression:', math.sqrt(mean_squared_error(y_train, y_pred_train_svr)))
print('Random Forest Regression:', math.sqrt(mean_squared_error(y_train, y_pred_train_rfr)))

Ridge Regression: 21777.719864704915
Lasso Regression: 21406.25485995607
Support Vector Machine Regression: 77824.86347394813
Random Forest Regression: 11123.154165904918


### RMSE_Validation

In [ ]:
print('Ridge Regression:', math.sqrt(mean_squared_error(y_val, y_pred_val_r)))
print('Lasso Regression:', math.sqrt(mean_squared_error(y_val, y_pred_val_l)))
print('Support Vector Machine Regression:', math.sqrt(mean_squared_error(y_val, y_pred_val_svr)))
print('Random Forest Regression:', math.sqrt(mean_squared_error(y_val, y_pred_val_rfr)))

Ridge Regression: 38361.607376978085
Lasso Regression: 38082.73213627771
Support Vector Machine Regression: 94329.36474310076
Random Forest Regression: 37507.572469761966


### In terms of RMSE, Random Forest Regression Model is the best model among models.

# Deployment

In [ ]:
import numpy as np
y_pred_test_rfr = rfr.predict(X_test)
output = pd.DataFrame({'Id': X_test.index, 'SalePrice': np.round(y_pred_test_rfr, 0)})
output.to_csv('submission.csv', index=False)

In [ ]:
pd.read_csv('submission.csv')

Id  SalePrice
0     1461   136938.0
1     1462   152120.0
2     1463   170148.0
3     1464   181021.0
4     1465   196863.0
...    ...        ...
1454  2915    87246.0
1455  2916    86751.0
1456  2917   164182.0
1457  2918   112871.0
1458  2919   233357.0

[1459 rows x 2 columns]